mmf/models/transformers/base.py

```diff
@@ -0,0 +1,204 @@
+        """Initialize the weights.
+        """
+        if isinstance(module, (nn.Linear, nn.Embedding)):
+            module.weight.data.normal_(mean=0.0, std=self.config.initializer_range)
+        elif isinstance(module, nn.module.LayerNorm):
+            module.bias.data.zero_()
+            module.weight.data.fill_(1.0)
+        if isinstance(module, nn.Linear) and module.bias is not None:
+            module.bias.data.zero_()
+
+    def tie_weights(self):
+        """Tie the weights between the input embeddings and the output embeddings
+        if required.
+        """
+        pass
+
+    def init_weights(self):
+        if self.config.random_initialize is False:
+            if self.config.transformer_base is None:
+                # No pretrained model, init weights
+                self.apply(self._init_weights)
+
+        # Tie weights if required
+        self.tie_weights()
+
+    def preprocess_sample(self, sample_list: Dict[str, Any]):
+        """Preprocess the sample_list. Generate image input ids, image position ids,
+        image mask. Create a attention mask for all the masks for different modalities.
+        """
+        # TODO(vedanuj): Make this method simpler
+        # Image feature(img_input_ids) and position(img_pos_ids) tokens
+        if "image_feature_0" in sample_list:
+            sample_list.img_input_ids = sample_list.image_feature_0
+            sample_list.img_pos_ids = sample_list.image_info_0.bbox
+        else:
+            sample_list.img_input_ids = self.vision_encoder(sample_list.image)
+            sample_list.img_pos_ids = sample_list.img_input_ids.new_tensor(
+                torch.arange(0, sample_list.img_input_ids.size(1), dtype=torch.long)
+                .unsqueeze(0)
+                .expand(
+                    sample_list.img_input_ids.size(0), sample_list.img_input_ids.size(1)
+                )
+                .unsqueeze(-1)
+            )
+
+        # Image mask
+        sample_list.image_mask = torch.ones_like(
+            sample_list.img_input_ids[:, :, 0], dtype=torch.long
+        )
+
+        # Attention mask
+        sample_list.attention_mask = torch.cat(
+            (sample_list.input_mask, sample_list.image_mask), dim=-1
+        )
+        sample_list.masked_lm_labels = getattr(sample_list, "lm_label_ids", None)
+        if sample_list.lm_label_ids is not None:
+            assert sample_list.masked_lm_labels.size(-1) == sample_list.input_mask.size(
+                -1
+            )
+            new_lm_labels = torch.ones_like(sample_list.attention_mask) * -1
+            size_masked_lm_labels = sample_list.masked_lm_labels.size()
+            assert len(size_masked_lm_labels) == 2
+            new_lm_labels[
+                : size_masked_lm_labels[0], : size_masked_lm_labels[1]
+            ] = sample_list.masked_lm_labels
+            sample_list.masked_lm_labels = new_lm_labels
+
+    def forward(self, sample_list: Dict[str, Any]):
```

Comment:
Document how would a SampleList in this case look like?

Reply:
Base class now holds no opinions how the sample list should look like.